### ISSUE TITLE
KBFS: Determine author in shared folders?

### ISSUE THREAD

**Issue by ibrado** [keybase/keybase-issues#2376]
Hi,

Not an issue, but a current stumper I hope someone can help me with:

Let's say we have `/keybase/private/user1,user2,user3`

Then someone drops a file in it. Is it possible to determine the author (or last editor) of that file?

Thanks!

**Comment by strib** [comment]
KBFS has the information, but doesn't currently expose it.  We'll probably have to wait until the Keybase GUI gets a little farther along, since there's no great ways to expose it just via the file system interface.  If you have suggestions, let me know!

**Comment by ibrado** [comment]
Actually, I just need the info via CLI... perhaps .kbfs_status_filename.ext ?

**Comment by strib** [comment]
Ah, yeah something like that could work.  Do you need it cryptographically verified that a certain person wrote the file?  The easiest thing for us to do, given the format of our data, is to give you: "this is who the most recent writer of the overall __claimed_ was the last person to write the file."  (Basically, you're trusting that the previous writer verified the previous signature, and so on back to the revision of the folder where the current version of the file first appeared.)

Obviously not as strong.  The stronger thing is possible, but would be a much longer road to get a good implementation.  We want to eventually, but it's unlikely to be a near-term priority.

**Comment by ibrado** [comment]
I think the "not as strong" solution would work great for most cases (at least until the stronger thing comes around), given that all the signing is done in the background anyway. I am not a crypto expert though...

Thanks a lot, and looking forward to this! :-)

**Comment by strib** [comment]
Cool, I filed an internal ticket for something like this.

**Comment by taruti** [comment]
@ibrado Do you need this typically for invidual files or all files in the directory?

i.e. as compared to `.kbfs_edit_history` in the folder?

**Comment by ibrado** [comment]
@taruti Any would do as long as I can parse it to get to a file's status... I can't find .kbfs_edit_history inside a `private` or `public` folder, btw. Perhaps this is a new development..?

```
cat: .kbfs_edit_history: No such file or directory
```

**Comment by strib** [comment]
@taruti: I think we should have it per-file; there's other info we might want in there.

`.kbfs_edit_history` may not be in a public build yet -- we're still working on making it useful.

**Comment by cjb** [comment]
It's in the latest public builds for Linux and Windows, I think -- the last public build for macOS is from 6 days ago.

**Comment by strib** [comment]
We now have a `.kbfs_fileinfo_$FILE` to find out what the client's view of the last writer of a file is.  For example:

```sh
$ cd /keybase/public/chris
$ ls
end-to-end.txt  keybase.txt            Photos    security_stuff
fruitpunch.txt  lemon_party.jpg        plan.txt  verified_software
index.md        Man in the Middle.psd  random    zcash.txt
$ cat .kbfs_fileinfo_lemon_party.jpg
{
  "LastWriterUnverified": "chris",
  "BlockInfo": {
    "ID": {},
    "KeyGen": -1,
    "DataVer": 1,
    "Creator": "23260c2ce19420f97b58d7d95b68ca00",
    "Writer": "",
    "RefNonce": [
      128,
      28,
      212,
      200,
      89,
      147,
      49,
      21
    ],
    "EncodedSize": 32825
  }
}
```

`LastWriterUnverified` is not cryptographically verified (hence the name); it can theoretically be spoofed by a valid writer to that folder if they hack their client.  We'll have better tools in the future with stronger guarantees, but this should get people going for now.

**Comment by strib** [comment]
(Also note that this won't work for files with really long names that are already close to the operating system limit on file name length.)

**Comment by ibrado** [comment]
Woohoo! Thanks so much!

**Comment by malgorithms** [comment]
In addition to large filenames, also worth noting this doesn't work for seeing who authored a sym link (unless I'm mistaken @strib )

**Comment by strib** [comment]
Yes, good point.  We actually don't even keep info on symlink creators right now.  If you think it's important, we can look into that.
